### PR TITLE
fix atomic forms and multimedia edge case issue

### PR DESCRIPTION
**Pull request by czue**
should fix http://manage.dimagi.com/default.asp?158373
- first commit adds a failing test (involves a small refactor) that reveals the issue
- second commit is the fix.

this is @dannyroberts territory, though @snopoke and @emord both looked into this and maybe could review while he's offline.

I'm not thrilled about the solution as it is a bit of a band-aid, but I think it's not horrible.

(wait for build)

### CODE REVIEW

**Comment by dimagimon** [comment]
:+1: No lint errors found.

File: corehq/ex-submodules/casexml/apps/case/models.py [hunk 1]
```diff
@@ -655,11 +655,12 @@ def _safe_replace_and_force_to_string(me, attr, val):
                                                               create_action)
         self.actions.append(create_action)
     
-    def update_from_case_update(self, case_update, xformdoc):
+    def update_from_case_update(self, case_update, xformdoc, other_forms=None):
```
Comment (czue) [inline corehq/ex-submodules/casexml/apps/case/models.py]: support passing more than just the primary form in memory - just had to push this down to get it to the call where it was needed.

File: corehq/ex-submodules/casexml/apps/case/models.py [hunk 2]
```diff
@@ -756,7 +757,9 @@ def _new_update_from_case_update(self, case_update, xformdoc):
                 )
                 self.actions.append(case_action)
 
-        self.rebuild(strict=False, xforms={xformdoc._id: xformdoc})
+        local_forms = {xformdoc._id: xformdoc}
+        local_forms.update(other_forms)
+        self.rebuild(strict=False, xforms=local_forms)
```
Comment (czue) [inline corehq/ex-submodules/casexml/apps/case/models.py]: down here.

File: corehq/ex-submodules/casexml/apps/case/tests/test_multimedia.py [hunk 3]
```diff
@@ -75,19 +77,24 @@ def _calc_file_hash(self, key):
...
         """
-        form = post_xform_to_couch(xml_data, dict_attachments)
-        form.domain = TEST_DOMAIN
+        sp = couchforms.SubmissionPost(
+            instance=xml_data,
+            domain=TEST_DOMAIN,
+            attachments=dict_attachments,
+            last_sync_token=sync_token,
+        )
+        response, xform, cases = sp.run()
```
Comment (czue) [inline corehq/ex-submodules/casexml/apps/case/tests/test_multimedia.py]: prior to this change the multimedia tests weren't exercising the atomic submissions code

**Comment by snopoke** [comment]
Wow, nice work finding this. Couple of test failures maybe related. Tried rebuild and will see if same tests fail.

**Comment by snopoke** [comment]
Failing test passes locally.

**Comment by czue** [comment]
kicked build again - doesn't look related but still good to wait

**Comment by snopoke** [comment]
Yea, build is set to fail fast so not all tests get run if there is a
failure

**Comment by benrudolph** [comment]
woohoo for tdd!

**Comment by czue** [comment]
(same test keeps failing so going to look into it)

**Comment by czue** [comment]
```
======================================================================
ERROR: test_selective_product_sync (corehq.apps.commtrack.tests.test_fixture.FixtureTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/dimagi/commcare-hq/corehq/apps/commtrack/tests/test_fixture.py", line 137, in test_selective_product_sync
    ).all(), key=lambda x: x.date)[-1]
TypeError: can't compare datetime.datetime to NoneType
----------------------------------------------------------------------
```

**Comment by dimagimon** [comment]
:+1: No lint errors found.

**Comment by dimagimon** [comment]
:+1: No lint errors found.

**Pull request merged**